### ISSUE TITLE
Failed sending HTTP/2 data to the peer

### ISSUE THREAD

**Issue by black-desk** [curl/curl#11859]
### I did this

I using ostree on debian sid to pull a lot of content from my ostree server behind a nginx.

ostree on debian using libcurl and I am sure that it using 8.2.1:

```
❯ ldd /usr/bin/ostree | grep curl
        libcurl-gnutls.so.4 => /lib/x86_64-linux-gnu/libcurl-gnutls.so.4 (0x00007fdbdfd44000)
```

After pulling some files, ostree failed with [55] Failed sending data to the peer.

I turn on CURLOPT_VERBOSE, then I got

```
* received GOAWAY, error=0, last_stream=1999
< HTTP/2 200
< date: Fri, 15 Sep 2023 07:05:58 GMT
< content-type: application/octet-stream
< content-length: 126
< accept-ranges: bytes
< last-modified: Tue, 11 Jan 2022 06:28:17 GMT
< strict-transport-security: max-age=15724800; includeSubDomains
<
* Closing connection

error: While fetching xxxxxx : [55] Failed sending data to the peer
```

Every time I failed with GOAWAY with error=0 and `last_stream=1999`.

Then I found this issue: https://github.com/golang/go/issues/18112

Basically, nginx has a default limitation about how many request can sent in a single http2 connection, it defaults to 1000.

When that limitation is reached, it send a GOAWAY with no error to tell the client side to create a new connection and close the old one.

### I expected the following

It seems that libcurl should retry when receive a GOAWAY without error.

### curl/libcurl version

8.2.1

### operating system

```
❯ uname -a
Linux black-desk-Debian-testing 6.1.0-10-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.38-1 (2023-07-14) x86_64 GNU/Linux
```

**Comment by black-desk** [comment]
I think #11496 maybe has the same reason. @calvin2021y

**Comment by bagder** [comment]
Any chance you can test 8.3.0 and see if the problem is still present or fixed?

**Comment by black-desk** [comment]
> Any chance you can test 8.3.0 and see if the problem is still present or fixed?

@bagder 

As debian and archlinux both have upgrade curl to 8.3.0, I have a try just now.

Unfortunately, this problem is still present.

**Comment by calvin2021y** [comment]
> I think #11496 maybe has the same reason. @calvin2021y

Yes, this problem cause me 30K firebase error report every day. I hope it fixed ASAP.

**Comment by calvin2021y** [comment]
confirm 8.3.0 still get this report from client.

```sh
55 0 Recv failure: Software caused connection abort Send failure: Broken pipe Recv failure: Software caused connection abort Send failure: Broken pipe POST
```

**Comment by Alphix** [comment]
I seem to have stumbled across the same bug, albeit using `git clone` instead (reported to the [git](https://lore.kernel.org/git/bb757ebd66b5ac4c81d62b01d5cff2f75250090d@hardeman.nu/T/#u) mailing list, where I was pointed in this direction).

**Comment by bagder** [comment]
Any chance we can get a stand-alone reproducible example source code? I don't know what we can do with this report without it.

**Comment by black-desk** [comment]
I think you should take a look on that golang issue I referenced above.

**Comment by black-desk** [comment]
I will post a example later.

**Comment by lukepalmer** [comment]
I also see this, and see that if I change my nginx setting "keepalive_requests" to be higher than the number of requests that I am making, the problem does not reproduce.

The nginx docs say that setting this "too high" is not recommended, whatever that means.
https://nginx.org/en/docs/http/ngx_http_core_module.html#keepalive_requests

Great job on the analysis @black-desk

**Comment by icing** [comment]
Any chance to reproduce this with a curl v8.3.0? Then you could add `-v --trace-config=ids,time,http/2` and we'd get a nice trace of what nginx is doing here and how curl fails the retry on another connection.

**Comment by lukepalmer** [comment]
Here we are! To keep it compact, this is with an nginx setting of "keepalive requests 3;" and making 5 requests. Thanks for looking.

[trace.txt](https://github.com/curl/curl/files/12837922/trace.txt)

**Comment by icing** [comment]
Thanks! Could you give #12054 a try?

**Comment by black-desk** [comment]
I just finish my example here: https://github.com/black-desk/curl-11859

**Comment by lukepalmer** [comment]
I re-ran my test cases with default nginx settings (reuse connections 1000 times) and #12054. I am now unable to reproduce this problem. Thanks to all who are involved!

Note that I can still cause failures that appear to be different with unrealistic nginx settings as detailed in #12054. My guess is that this is a different problem.

**Comment by bagder** [comment]
> Note that I can still cause failures that appear to be different with unrealistic nginx settings

If this is still an actual problem for you, please submit a new issue with all the details and we will work on that. I consider this original bug handled now with #12054 merged.
